src/tiledbsoma_ml/io_batches.py

```diff
@@ -0,0 +1,139 @@
+# Copyright (c) 2021-2024 The Chan Zuckerberg Initiative Foundation
+# Copyright (c) 2021-2024 TileDB, Inc.
+#
+# Licensed under the MIT License.
+
+import gc
+import logging
+import time
+from typing import Iterable, Iterator, Optional, Sequence, Tuple
+
+import attrs
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+from tiledbsoma import DataFrame, IntIndexer, SparseNDArray
+
+from tiledbsoma_ml._csr import CSR_IO_Buffer
+from tiledbsoma_ml._utils import EagerIterator, batched
+from tiledbsoma_ml.common import NDArrayJoinId
+from tiledbsoma_ml.query_ids import Chunks
+
+logger = logging.getLogger("tiledbsoma_ml.io_batches")
+IOBatch = Tuple[CSR_IO_Buffer, pd.DataFrame]
+"""Tuple type emitted by |IOBatches|, containing ``X`` rows (as a |CSR_IO_Buffer|) and ``obs`` rows (|pd.DataFrame|)."""
+
+
+@attrs.define(frozen=True)
+class IOBatches(Iterable[IOBatch]):
+    """Given a list of ``obs_joinid`` |Chunks|, re-chunk them into (optionally shuffled) |IOBatch|'s".
+
+    An |IOBatch| is a tuple consisting of a batch of rows from the ``X`` |SparseNDArray|, as well as the corresponding
+    rows from the ``obs`` |DataFrame|. The ``X`` rows are returned in an optimized |CSR_IO_Buffer|.
+    """
+
+    chunks: Chunks
+    io_batch_size: int
+    obs: DataFrame
+    var_joinids: NDArrayJoinId
+    X: SparseNDArray
+    obs_column_names: Sequence[str] = ("soma_joinid",)
+    seed: Optional[int] = None
+    shuffle: bool = True
+    use_eager_fetch: bool = True
+
+    @property
+    def io_batch_ids(self) -> Iterable[Tuple[int, ...]]:
+        """Re-chunk ``obs_joinids`` according to the desired ``io_batch_size``."""
+        return batched(
+            (joinid for chunk in self.chunks for joinid in chunk),
+            self.io_batch_size,
+        )
+
+    def __iter__(self) -> Iterator[IOBatch]:
+        """Emit |IOBatch|'s."""
+        # Create RNG - does not need to be identical across processes, but use the seed anyway
+        # for reproducibility.
+        shuffle_rng = np.random.default_rng(self.seed)
+        X = self.X
+        context = X.context
+        obs_column_names = (
+            list(self.obs_column_names)
+            if "soma_joinid" in self.obs_column_names
+            else ["soma_joinid", *self.obs_column_names]
+        )
+        # Round-trip though tuple avoids `TypeError: IntIndexer only supports array of type int64`.
+        # TODO: debug / work around that error; serde'ing the ndarray apparently results in a second np.int64 instance, that fails reference equality check vs. the version from the worker-process.
+        var_joinids = np.array(tuple(self.var_joinids))
```

Comment:
IntIndexer supports np.ndarray and Arrow integer arrays, as long as they are int64. Not sure why this is needed, but it will be quite expensive for large arrays.

I'd like to see this root caused in case it is actually an IntIndexer bug.  Can you provide any info on the actual type (both container and dtype) of the object?

it would also be much better to do a zero-copy cast if possible.

Reply:
BTW, the types (and their casts) supported by IntIndexer are mostly done in the Python code (in tiledbsoma), so we can easily extend that if we have missed something useful.

Reply:
This issue is still open, and is costing a full copy of the var ids.  I left a note in the new code to the same effect.

Reply:
Sorry to make you flag this twice, I believe I resolved the 2nd incarnation [below](https://github.com/single-cell-data/TileDB-SOMA-ML/pull/27#discussion_r1978581157).